pandas/tests/scalar/test_timestamp.py

```diff
@@ -175,6 +175,21 @@ def test_constructor_invalid(self):
         with tm.assert_raises_regex(ValueError, 'Cannot convert Period'):
             Timestamp(Period('1000-01-01'))
 
+    def test_constructor_invalid_tz(self):
+        # GH#17690, GH#5168
+        with tm.assert_raises_regex(TypeError, 'must be a datetime.tzinfo'):
+            Timestamp('2017-10-22', tzinfo='US/Eastern')
+
+        with tm.assert_raises_regex(ValueError, 'at most one of'):
+            Timestamp('2017-10-22', tzinfo=utc, tz='UTC')
+
+        with tm.assert_raises_regex(ValueError, 'Use `tzinfo` instead'):
+            Timestamp(year=2017, month=10, day=22, tz='UTC')
+
+        with tm.assert_raises_regex(ValueError, 'Use `tz` instead'):
+            dt = datetime(2017, 10, 22)
+            Timestamp(dt, tzinfo=utc)
```

Comment:
this is a valid case. you can just do (internally).

```
tz, tzinfo = tzinfo, None
```

Reply:
Hmm this gets really verbose since we need to check in a handful of different places.  I guess get correctness right first, then circle back for brevity.